Monika After Story/game/script-affection.rpy

```diff
@@ -1675,19 +1682,50 @@ init 20 python:
         # Updates the experience levels if necessary.
         mas_updateAffectionExp()
 
-    def mas_setApologyReason(reason):
+    def mas_setApologyReason(
+        reason=None,
+        ev_label=None,
+        apology_active_expiry=datetime.timedelta(hours=3),
+        apology_overall_expiry=datetime.timedelta(weeks=1)
+        ):
         """
         Sets a reason for apologizing
 
         IN:
-            reason - The reason for the apology
+            reason - The reason for the apology (integer value corresponding to item in the apology_reason_db)
+                (if left None, and an ev_label is present, we assume a non-generic apology)
+            ev_label - The apology event we want to unlock
                 (required)
-
+            apology_active_expiry - The amount of session time after which, the apology that was added expires
+                defaults to 3 hours active time
+            apology_overall_expiry - The amount of overall time after which, the apology that was added expires
+                defaults to 7 days
         """
+
         global mas_apology_reason
 
-        mas_apology_reason = reason
+        if ev_label is None:
+            if reason is None:
+                mas_apology_reason = 0
+            else:
+                mas_apology_reason = reason
+            return
+        elif mas_getEV(ev_label) is None:
+            store.mas_utils.writelog(
+                "[ERROR]: ev_label does not exist: {0}\n".format(repr(ev_label))
+            )
+            return
 
+        if ev_label not in persistent._mas_apology_time_db:
+            #Unlock the apology ev label
+            store.mas_showEVL(ev_label, 'APL',unlock=True)
+
+            #Calculate the current total playtime
+            current_total_playtime = persistent.sessions['total_playtime'] + (datetime.datetime.now() - persistent.sessions['current_session_start'])
```

Comment:
don't we have a function somewhere for getting current session time? or was that in a pull request that is still under review?
If the latter, can you mark a TODO here for that.